### PR TITLE
Add option for class specific filtering.

### PR DESCRIPTION
**Pull request by hgaiser**
This PR adds a flag to disable class specific filtering. When disabled, each anchor would take the maximum scoring class and use that, instead of using all classes that have a score above a certain threshold.

### CODE REVIEW

**Comment by netw0rkf10w** [comment]
I have tried this and it seems that NMS does not work properfly.

What I did is replacing the old code with the new one (downloaded from the [`class-specific-filtering` branch](https://github.com/fizyr/keras-retinanet/tree/class-specific-filtering)), reinstalling it (`pip install .`, because I'm using a virtual environment), then launching the training script from scratch. After 3 epochs I converted the model using 

`keras_retinanet/bin/convert_model.py --backbone resnet101 --no-class-specific-filter ./snapshots/resnet101_csv_03.h5 ./models/resnet101_csv_inference_03.h5`

and then used the converted model in my inference code.

Some results can be seen here: https://github.com/fizyr/keras-retinanet/issues/466#issuecomment-399096818

I have checked the values in `filter_detections.py` (two instances of them) and they are correct:

```
        nms                   = True,
        class_specific_filter = True,
        nms_threshold         = 0.5,
        score_threshold       = 0.05,
```

**Comment by hgaiser** [comment]
So to be clear @netw0rkf10w , did  you use this branch, or did you copy parts of the code to master?

**Comment by netw0rkf10w** [comment]
@hgaiser What I did is:
- Rename the folder containing the old code from `keras-retinanet` to `keras-retinanet_OLD`.
- Download the whole branch [class-specific-filtering](https://github.com/fizyr/keras-retinanet/tree/class-specific-filtering) and extract it to the same place as previous code: `keras-retinanet`.
- Go to `keras-retinanet` and `pip install .`.

Using the above steps I think I'm using the new code now.

**Comment by hgaiser** [comment]
If you used `--no-class-specific-filter` argument, then yes you should be using the branch (otherwise it would say `--no-class-specific-filter` is an unknown argument).

**Comment by netw0rkf10w** [comment]
@hgaiser Sorry I didn't know that you replied (I thought the notification was for the previous reply). Yes I used `--no-class-specific-filter` without errors.

**Comment by netw0rkf10w** [comment]
@hgaiser To be sure I even trained my model from scratch with the new code.
Oh, I didn't know that you guys have an implementation of Mask R-CNN!!! That's awesome! I have been using matterport's one, but I'm going to test yours very soon! Thank you so much!

**Comment by enricoliscio** [comment]
LGTM, you can merge

**Comment by hgaiser** [comment]
> @hgaiser To be sure I even trained my model from scratch with the new code.
> Oh, I didn't know that you guys have an implementation of Mask R-CNN!!! That's awesome! I have been using matterport's one, but I'm going to test yours very soon! Thank you so much!

It's not advertised, but it's there and it's working :)

It's technically not really Mask-RCNN though. It's a combination of MaskRCNN and RetinaNet (see the disclaimer on keras-maskrcnn).

**Comment by raeony** [comment]
@hgaiser @netw0rkf10w 
firstly, thanks for the great work you did!
I have tried the new codes you two discussed in my datasets, acctually the NMS didn't work still(the old codes didn't work yet). so I want to know what's  problem with this , do you have a solution to solve this?
the overlapping boxes are just like this:
image

**Comment by netw0rkf10w** [comment]
@raeony Your results look strange. Which code are you using? Did you uninstall the previous version before installing the new one?
